types/wlr_buffer.c

```diff
@@ -330,3 +362,88 @@ struct wlr_client_buffer *wlr_client_buffer_apply_damage(
 	buffer->resource_released = true;
 	return buffer;
 }
+
+static const struct wlr_buffer_impl shm_client_buffer_impl;
+
+static struct wlr_shm_client_buffer *shm_client_buffer_from_buffer(
+		struct wlr_buffer *buffer) {
+	assert(buffer->impl == &shm_client_buffer_impl);
+	return (struct wlr_shm_client_buffer *)buffer;
+}
+
+static void shm_client_buffer_destroy(struct wlr_buffer *wlr_buffer) {
+	struct wlr_shm_client_buffer *buffer =
+		shm_client_buffer_from_buffer(wlr_buffer);
+	wl_list_remove(&buffer->resource_destroy.link);
+	wl_list_remove(&buffer->release.link);
+	free(buffer);
+}
+
+static bool shm_client_buffer_begin_data_ptr_access(struct wlr_buffer *wlr_buffer,
+		void **data, uint32_t *format, size_t *stride) {
+	struct wlr_shm_client_buffer *buffer =
+		shm_client_buffer_from_buffer(wlr_buffer);
+	enum wl_shm_format wl_shm_format =
+		wl_shm_buffer_get_format(buffer->shm_buffer);
+	// TODO: protect access between begin_access and end_access
+	*data = wl_shm_buffer_get_data(buffer->shm_buffer);
+	*format = convert_wl_shm_format_to_drm(wl_shm_format);
+	*stride = wl_shm_buffer_get_stride(buffer->shm_buffer);
+	wl_shm_buffer_begin_access(buffer->shm_buffer);
+	return true;
+}
+
+static void shm_client_buffer_end_data_ptr_access(struct wlr_buffer *wlr_buffer) {
+	struct wlr_shm_client_buffer *buffer =
+		shm_client_buffer_from_buffer(wlr_buffer);
+	wl_shm_buffer_end_access(buffer->shm_buffer);
+}
+
+static const struct wlr_buffer_impl shm_client_buffer_impl = {
+	.destroy = shm_client_buffer_destroy,
+	.begin_data_ptr_access = shm_client_buffer_begin_data_ptr_access,
+	.end_data_ptr_access = shm_client_buffer_end_data_ptr_access,
+};
+
+static void shm_client_buffer_resource_handle_destroy(
+		struct wl_listener *listener, void *data) {
+	struct wlr_shm_client_buffer *buffer =
+		wl_container_of(listener, buffer, resource_destroy);
+	buffer->resource = NULL;
```

Comment:
The `buffer->shm_buffer` is destroyed with the resource. We might still be able to access the underlying data if we ref the shm pool and save all of the shm buffer params. But that's a little awkward, maybe it would be nicer to improve the libwayland API instead.

Reply:
Fixed.